src/local/butler/reproduce_tool/android.py

```diff
@@ -0,0 +1,112 @@
+EMULATOR_PATH = os.path.join(
+    os.path.dirname(__file__), os.pardir, os.pardir, os.pardir, os.pardir,
+    'local', 'bin', 'android-sdk', 'emulator', 'emulator')
+
+
+def start_emulator():
+  """Return a ProcessRunner configured to start the Android emulator."""
+  runner = new_process.ProcessRunner(
+      EMULATOR_PATH,
+      ['-avd', 'Android28', '-writable-system', '-partition-size', '2048'])
+  emulator_process = runner.run()
+
+  # If we run adb commands too soon after the emulator starts, we may see
+  # flake or errors. Delay a short while to account for this.
+  # TODO(mbarbella): This is slow and flaky, but wait-for-device isn't usable if
+  # another device is connected (as we don't know the serial yet). Find a better
+  # solution.
+  time.sleep(30)
+
+  return emulator_process
+
+
+def get_devices():
+  """Get a list of all connected Android devices."""
+  adb_runner = new_process.ProcessRunner(adb.get_adb_path())
+  adb_runner.run_and_wait(additional_args=['run-server'])
+  result = adb_runner.run_and_wait(additional_args=['devices'])
+
+  if result.return_code:
+    raise errors.ReproduceToolUnrecoverableError('Unable to run adb.')
+
+  # Ignore non-device lines (those before "List of devices attached").
+  store_devices = False
+  devices = []
+  for line in result.output.splitlines():
+    if line == ADB_DEVICES_SEPARATOR_STRING:
+      store_devices = True
+      continue
+    if not store_devices or not line:
+      continue
+
+    devices.append(line.split()[0])
+
+  return devices
+
+
+def prepare_environment(disable_android_setup):
+  """Additional environment overrides needed to run on an Android device."""
+  environment.set_value('OS_OVERRIDE', 'ANDROID')
+
+  # Bail out if we can't determine which Android device to use.
+  serial = environment.get_value('ANDROID_SERIAL')
+  if not serial:
+    devices = get_devices()
+    if len(devices) == 1:
+      serial = devices[0]
+      environment.set_value('ANDROID_SERIAL', serial)
+    elif not devices:
+      raise errors.ReproduceToolUnrecoverableError(
+          'No connected Android devices were detected. Run with the -e '
+          'argument to use an emulator.')
+    else:
+      raise errors.ReproduceToolUnrecoverableError(
+          'You have multiple Android devices or emulators connected. Please '
+          'set the ANDROID_SERIAL environment variable and try again.\n\n'
+          'Attached devices: ' + ', '.join(devices))
+
+  print('Warning: this tool will make changes to settings on the connected '
+        'Android device with serial {serial} that could result in data '
+        'loss.'.format(serial=serial))
+  willing_to_continue = prompts.get_boolean(
+      'Are you sure you want to continue?')
+  if not willing_to_continue:
+    raise errors.ReproduceToolUnrecoverableError(
+        'Bailing out to avoid changing settings on the connected device.')
+
+  if 'emulator' in serial:
+    adb.run_command(['remount'])
```

Comment:
could you please add a comment here explaining why this is only needed for the emulator case?

Reply:
This is actually unnecessary, so I've removed it. It's a leftover from an issue I was debugging.